webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/NodeFactQueryProcessor.scala

```diff
@@ -112,7 +113,22 @@ object GroupOr extends Group {
   }
   def inverse(a: NodeFactMatcher):                     NodeFactMatcher =
     NodeFactMatcher(s"!(${a.debugString})", (n: CoreNodeFact) => a.matches(n).map(!_))
-  val zero:                                            NodeFactMatcher = NodeFactMatcher("false", _ => false.succeed)
+  val zero:                                            NodeFactMatcher = NodeFactMatcher("or0_false", _ => false.succeed)
+}
+
+/*
+ * A case class to sort a list of criterion lines into the different kind we know about:
+ * - core node fact
+ * - sub group
+ * - ldap (historical)
+ */
+final case class CriterionLines(
+    coreNodeFact: Chunk[CriterionLine],
+    subGroup:     Chunk[CriterionLine],
+    ldap:         Chunk[CriterionLine]
+)
```

Comment:
That case class and the following lines correct one the big problem, ie the fact that we weren't really splitting between matchers kind, but just collecting several times, creating duplicates along the way. 
Now we cleanly check each line in turn and put in one (and only one) of the matcher kind and then create the matcher for its lines.